src/main/scala/firrtl/Utils.scala

```diff
@@ -12,11 +12,55 @@ import scala.collection.mutable.{StringBuilder, ArrayBuffer, LinkedHashMap, Hash
 import java.io.PrintWriter
 import logger.LazyLogging
 
-class FIRRTLException(str: String) extends Exception(str)
+object FIRRTLException {
+  def defaultMessage(message: String, cause: Throwable) = {
+    if (message != null) {
+      message
+    } else if (cause != null) {
+      cause.toString
+    } else {
+      null
+    }
+  }
+}
+class FIRRTLException(str: String, cause: Throwable = null)
+  extends RuntimeException(FIRRTLException.defaultMessage(str, cause), cause)
```

Comment:
Here's the motivation:
[extending-exception-runtimeexception-in-java](http://stackoverflow.com/questions/19857008/extending-exception-runtimeexception-in-java), [runtimeexception controversy](https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html)

Ideally we should have two categories of FIRRTL exceptions/errors:
- parsing, input errors
- programming - "can't happen here" errors

and we should throw a different exception for each of those conditions.

The "parsing, input errors" can be addressed (at least potentially), by the client, either by changing the input or environment. Programming errors can only be addressed by a developer and require updating firrtl source code.

I'm not sure if it's clear which of those categories a MatchError falls into.

The nulls are inherited from Java.